File: docs/greedy/0768_hard/max_chunks_stack_version.py

def find_max_sortable_chunks(nums: list[int]) -> int:  # LeetCode Q.768 & 769.
    stack: list[int] = []

    for num in nums:
        if not stack:
            stack.append(num)
            continue

        chunk_max = stack[-1]
        
        while stack and stack[-1] > num:
            stack.pop(-1)
        
        stack.append(max(chunk_max, num))

    return len(stack)

File: docs/greedy/0768_hard/max_chunks_stack_version.cpp
#include <vector>
using namespace std;
#include <stack>

int findMaxSortableChunks(vector<int>& nums) // LeetCode Q.768 & 769.
{
    stack<int> stack;

    for (const auto& num : nums) {
        if (stack.empty()) {
            stack.push(num);
            continue;
        }

        int chunkMax = stack.top();

        while (!stack.empty() && num < stack.top())
            stack.pop();

        stack.push(max(chunkMax, num));
    }

    return stack.size();
}
